src/cargo/ops/cargo_compile.rs

```diff
@@ -944,6 +945,48 @@ fn generate_targets<'a>(
     Ok(units.into_iter().collect())
 }
 
+fn warn_on_missing_features(resolve: &Resolve, required_features: &[String], summary: &Summary) {
+    for feature in required_features {
+        match FeatureValue::new(feature.into(), summary) {
+            // No need to do anything here, since the feature must exist to be parsed as such
+            FeatureValue::Feature(_) => {}
+            // Possibly mislabeled feature that was not found
+            FeatureValue::Crate(krate) => {
+                if summary
+                    .dependencies()
+                    .iter()
+                    .map(|d| d.name_in_toml())
+                    .find(|name| name == &krate)
+                    .is_none()
```

Comment:
I would probably not separate these, to keep it simpler, and use `any`.  I think it should also check that it is optional.  Maybe something like this:

```suggestion
                if !summary
                    .dependencies()
                    .iter()
                    .any(|d| d.name_in_toml() == krate && d.is_optional())
```